Documentation/design-docs/multi-reg-call-nodes.md

```diff
@@ -0,0 +1,211 @@
+
+Arm32:
+Long type value is returned in two return registers r0 and r1
+HFA structs will be returned in 1-4 successive VFP registers s0-s3 or d0-d3
+
+x86:
+Long type value is returned in two return registers EDX and EAX.
+
+Post Importation IR Forms:
+-------------------------
+In importer any call returning a (struct or long type) value in two or more registers is forced to a temp
+and temp is used in place of call node subsequently in IR. Note that creation of 'tmp' is avoided if return 
+value of call is getting assigned to a local in IL.
+
+// tmp = GT_CALL, where tmp is of TYP_STRUCT or TYP_LONG
+GT_ASG(TYP_STRUCT or TYP_LONG, tmp, GT_CALL node)
+
+Similarly importer will force IR of GT_RETURN node returning a value in multiple return registers to be
+of the following form if operand of GT_RETURN is anything but an lcl var
+
+GT_ASG(TYP_STRUCT or TYP_LONG, tmp, op1)
+GT_RETURN(tmp)
+
+Post struct promotion IR forms:
+------------------------------
+Before global morph of basis blocks, struct promotion takes place. It will give rise to the following
+three cases:  
+
+Case 1:
+tmp is not struct promoted or Type Dependently Promoted (P-DEP).
+
+Case 2:
+tmp is Type Independently Promoted (P-INDEP) but its field layout doesn't match the layout of return registers or
+tmp is P-FULL promoted struct (e.g. SIMD type structs).
+
+For example, tmp could be a struct of 4 integers.  But on x64 unix, two fields of such a struct
+will be returned in one 8-byte return register.
+
+Case 3:
+tmp is P-INDEP promoted and its field layout matches the layout of return registers.
+An example is a struct containing two fields of {TYP_REF, TYP_DOUBLE}.
+
+Post Global Morph, IR forms of tmp=GT_CALL where tmp is of TYP_STRUCT
+---------------------------------------------------------------------
+Case 3 is morphed into
+
+   GT_STORE_MULTI_VAR(TYP_STRUCT, <FieldLcl1, FieldLcl2, FieldLcl3, FieldLcl4>, op1 = GT_CALL)
+
+   Where FieldLcl[1..4] are lcl numbers of P-INDEP promoted fields of tmp.
+
+   GT_STORE_MULTI_VAR is a new GenTree node to represent the store operation to 2-4 locals
+   using multi-reg/mem value of a call/lclvar respectively.  It also will have additional fields
+   to store the registers into which FieldLcl[1..4] need to be stored and also a spill mask
+   to indicate which of the locals needs to be spilled. 
+
+Case 1 is morphed as    
+   GT_OBJ(&tmp) = GT_CALL
+
+   Post rationalization this will get converted to GT_STORE_OBJ/BLK(&tmp, GT_CALL) and 
+   block op codegen will special case for multi-reg call case.
+
+Case 2 can be handled one of the following two ways
+   a) Force tmp to memory and morph it as in case 1 above
+   b) Create 2-4 temp locals matching the type of respective return registers of GT_CALL and 
+      use them to create following IR
+
+      GT_STORE_MULTI_VAR(TYP_STRUCT, <tmpLcl1, tmpLcl2, tmpLcl3, tmpLcl4>, op1 = GT_CALL)
+      IR nodes to extract/assemble fields of tmp struct from individual tmpLcl[1..4]
```

Comment:
You might want to have an example here, e.g. a struct on amd64 that has two 32-bit integer fields, and one TYP_REF field - and show what the IR would look like (the GT_STORE_MULTI_VAR followed by 2 or 3 assignments (one could possibly optimize due to the fact that the TYP_REF field would be a match).